model/kubernetes/kubernetes_cluster.rb

```diff
@@ -60,6 +71,93 @@ def self.kubeconfig(vm)
   def kubeconfig
     self.class.kubeconfig(cp_vms.first)
   end
+
+  def add_finalizer(svc)
+    kubectl("-n #{svc["metadata"]["namespace"]} patch service #{svc["metadata"]["name"]} --type='merge' -p '{\"metadata\":{\"finalizers\":[\"#{KubernetesCluster::SERVICE_FINALIZER}\"]}}'")
+  end
+
+  def remove_finalizer(svc)
+    kubectl("-n #{svc["metadata"]["namespace"]} patch service #{svc["metadata"]["name"]} --type='json' -p='[{\"op\": \"remove\", \"path\": \"/metadata/finalizers\",\"value\":[\"#{KubernetesCluster::SERVICE_FINALIZER}\"]}]'")
+  end
+
+  def set_load_balancer_host_name(svc, hostname)
+    kubectl("-n #{svc["metadata"]["namespace"]} patch service #{svc["metadata"]["name"]} --type='merge' -p '{\"status\":{\"loadBalancer\":{\"ingress\":[{\"hostname\":\"#{hostname}\"}]}}}' --subresource='status'")
+  end
+
+  def lb_name_for_svc
+    ubid
+  end
+
+  # Returning an array in this format:
+  # [[src_port_0, dst_port_0], [src_port_1, dst_port_1],...]
+  def lb_desired_ports(svc)
+    svc["spec"]["ports"].map { |port| [port["port"], port["nodePort"]] }
+  end
+
+  def is_deleted(svc)
+    svc.dig("metadata", "deletionTimestamp")
+  end
+
+  def extra_and_missing_vms_for_lb(lb)
+    worker_vms = nodepools.flat_map(&:vms)
+    worker_vm_ids = worker_vms.map(&:id).to_set
+    lb_vms = lb.load_balancers_vms.map(&:vm)
+    lb_vm_ids = lb_vms.map(&:id).to_set
+
+    extra_vms = lb_vms.reject { |vm| worker_vm_ids.include?(vm.id) }
+    missing_vms = worker_vms.reject { |vm| lb_vm_ids.include?(vm.id) }
+    [extra_vms, missing_vms]
+  end
+
+  def extra_and_missing_ports_for_lb(lb, desired_ports)
+    lb_ports_hash = lb.ports.to_h { |p| [[p.src_port, p.dst_port], p.id] }
+    missing_ports = desired_ports - lb_ports_hash.keys
+    extra_ports = (lb_ports_hash.keys - desired_ports).map { |p| LoadBalancerPort[id: lb_ports_hash[p]] }
+
+    [extra_ports, missing_ports]
+  end
+
+  def detect_loadbalancer_svc_updates(svc)
+    return true if is_deleted(svc) && svc.dig("metadata", "finalizers")&.include?(SERVICE_FINALIZER)
+
+    lb = LoadBalancer.where(name: lb_name_for_svc).first
+    return true if lb.nil?
+
+    extra_vms, missing_vms = extra_and_missing_vms_for_lb(lb)
+    extra_ports, missing_ports = extra_and_missing_ports_for_lb(lb, lb_desired_ports(svc))
+
+    hostname_set_on_status = !svc.dig("status", "loadBalancer", "ingress", "hostname").to_s.empty?
+
+    !extra_vms.empty? || !missing_vms.empty? || !extra_ports.empty? || !missing_ports.empty? || !hostname_set_on_status
```

Comment:
instead of chaining clauses, you can exit early at the time of detecting some change.